### ISSUE TITLE
research: Spikenaut supervisor v3 — shadow policy, online adaptation, and hardware parity

### ISSUE THREAD

**Issue by rmems** [rmems/Spikenaut-SNN#6]
## Execution status — reconciled 2026-09-19

Reopened: PR #57 delivered [RM-1328](https://linear.app/rpd-34/issue/RM-1328/featpolicy-define-deterministic-spikenaut-output-row-decision-contract)'s output-row decision contract and incorrectly closed this entire research program. PR #56 delivered model-bank attestation. Neither establishes the parent acceptance criteria.

### Next deliverable: frozen software shadow replay

- [X] Attested model-bank selection (PR #56 / [RM-1327](https://linear.app/rpd-34/issue/RM-1327/featartifacts-attest-spikenaut-model-bank-entries-before-selection)).
- [X] Deterministic output-row conversion (PR #57 / [RM-1328](https://linear.app/rpd-34/issue/RM-1328/featpolicy-define-deterministic-spikenaut-output-row-decision-contract)).
- [X] Analog five-sensor front end (PR #54) and checked signed export (PR #58).
- [ ] One command assembles attested checkpoint → LiveStimAdapter-compatible inputs → reference LIF stepping → readout → decision diagnostics.
- [ ] Save checkpoint/dataset hashes, session splits, train-only normalization, seed/configuration, per-step outputs and repeatable summary; reject overlap and invalid telemetry explicitly.
- [ ] Evaluate the actual comfort/temp/power continuous targets against persistence, linear and memoryless baselines on held-out sessions (#21).
- [ ] Define and validate any action policy separately. The three regression outputs do not establish ALLOW/WARN/THROTTLE/PAUSE/YIELD_GPU; never interpret their argmax as a validated safety action.
- [ ] Bind software/RTL/board replay to an explicit input modality and timing contract; retain rmems/silicon-hdl#76 as hardware owner.

Current focus is M0 replay qualification → M1 baseline evidence. M4 parity preparation can proceed alongside frozen evaluation. M3 adaptive promotion follows a frozen baseline and compatible checkpoint handoff (#17); M5/M6 remain deferred.

The historical feature suggestions below are superseded by the workload-agnostic five-sensor exp-025 contract. Workload identity is a label, not a model input; missing cognitive signals are never fabricated. The synthetic neuromod host experiment does not run the shipped signed bank.

Parent: #7. This issue remains the cross-stage research umbrella; assigning M1 indicates the current deliverable, not completion of later stages.

---

## Planning attribution

**GPT-5.6 (high)**

## Context

`Spikenaut-SNN` has already been used as a telemetry-trained neuromorphic model and has historical supervisor ambitions, but prior FPGA operation was not instrumented strongly enough to answer a basic question: **did the deployed SNN actually behave like the software model, and did its decisions help?**

This issue defines the next supervisor experiment as a measured research program rather than immediately putting an adaptive model back in control of the workstation.

Existing relevant work:

* #2 — degenerate/monotonic hidden-weight root cause and data-quality findings
* #3 — inhibitory connections + recurrence roadmap
* #4 — Q8.8 export fidelity investigation
* `rmems/Theseus-Quarry` — deterministic GPU/process safety layer + operational telemetry
* `rmems/gaming-telemetry` — high-frequency machine telemetry
* `rmems/SynapticDistill.jl` — online/teacher-student SNN training
* `rmems/silicon-bridge` / `rmems/silicon-hdl` — host↔FPGA deployment path

## Research hypothesis

A small event-driven SNN can act as a useful **interoceptive supervisor** by compressing continuous machine telemetry into a compact latent/control state and learning bounded resource-management decisions. The SNN should be treated as an adaptive policy layer, while deterministic Rust safety rules remain authoritative.

## Goal

Produce **Spikenaut Supervisor v3** with three independently testable stages:

1. software shadow supervisor
2. bounded online adaptation
3. software↔FPGA parity evaluation

Do not enable unrestricted host control as part of this issue.

## Stage 0 — prerequisites / data readiness

Before training a new supervisor checkpoint:

- [ ] resolve or explicitly scope remaining dataset-quality blockers from #2/#3
- [ ] define the canonical input feature map and missing-value behavior
- [ ] define train/validation/test **session splits**
- [ ] record normalization statistics from train split only
- [ ] include dataset manifest/hash with every trained checkpoint

Recommended initial feature classes:

```text
thermal: gpu_temp, cpu_temp
energy: gpu_power, cpu_power
capacity: vram_used, gpu_util, cpu_util
transport: pcie rx/tx rate
status: throttle flags / pstate
workload: coarse workload class or one-hot label
```

Do not force exactly 16 raw fields merely because the current model has 16 input channels; add an explicit encoder/adapter if the research representation changes.

## Stage 1 — software shadow supervisor

Define a small output policy head suitable for replay evaluation.

Suggested action vocabulary:

```text
ALLOW
WARN
THROTTLE
PAUSE
YIELD_GPU
```

The exact vocabulary must match the consuming experiment and should stay smaller than the raw action space at first.

Requirements:

- [ ] software reference implementation produces deterministic outputs with fixed seed/checkpoint
- [ ] outputs include logits/scores or confidence, not only hard labels
- [ ] replay harness logs model state/output at each decision step
- [ ] no process control from the model in Stage 1

### Required baselines

Compare against:

* deterministic threshold policy
* simple memoryless baseline (linear/MLP)
* optional larger teacher (`LiquidCortex` or another frozen policy)

Metrics:

* macro F1 / confusion matrix
* false-safe rate (`ALLOW` on a safety-critical target)
* warning lead time
* state forecast MAE if continuous head exists
* spike rate
* inference latency
* parameter/memory footprint

## Stage 2 — bounded online adaptation

After an offline checkpoint is reproducible, add an opt-in online learning mode.

### Two-clock rule

Keep inference and learning conceptually separate:

```text
FAST LOOP
telemetry → spikes → inference → proposal

SLOW LOOP
outcome/reward → eligibility update → bounded parameter update
```

Do not perform unconstrained weight changes every raw sensor sample.

Requirements:

- [ ] configurable update cadence
- [ ] checkpoint before/after each online session
- [ ] maximum update magnitude / clipping or another explicit stability bound
- [ ] ability to replay the same session with online learning disabled
- [ ] log number and magnitude of updates
- [ ] detect NaN/overflow/out-of-range parameters and fail closed

### Reward components

Keep raw reward components separable, e.g.:

* thermal headroom
* VRAM headroom
* time throttled
* task/workload throughput
* unnecessary intervention penalty
* action-churn penalty

Do not hide everything behind one undocumented scalar.

## Stage 3 — FPGA parity

The first FPGA goal is **parity, not autonomy**.

For identical encoded input sequences, compare software Spikenaut against the FPGA implementation.

Required parity metrics:

* action agreement rate
* spike bitmap agreement
* membrane-potential error where readable
* Q8.8 quantization error
* decision latency
* false-safe disagreement count
* reproducibility across repeated board runs

Store a golden replay bundle containing:

```text
input sequence
software spikes/state
software action/logits
expected Q8.8 parameters
FPGA readback
parity summary
```

If board state cannot expose membrane potentials yet, record that limitation and start with spike/action parity.

## Safety boundary

Spikenaut may eventually **propose** actions to `Theseus-Quarry`, but:

- [ ] it cannot override thermal emergency
- [ ] it cannot raise safety thresholds
- [ ] invalid output maps to deterministic fallback / no-op
- [ ] model crash must not disable telemetry or Rust governor
- [ ] assisted control is a separate opt-in milestone after shadow metrics exist

## Suggested artifact/version naming

```text
spikenaut-supervisor-v3-exp001
spikenaut-supervisor-v3-exp002
...
```

Each experiment manifest should include:

* training commit
* dataset manifest/hash
* feature map
* encoder config
* learning rule
* seed
* checkpoint hash
* evaluation session ids
* software/FPGA result hashes where applicable

## Tests / validation

- [ ] fixed checkpoint + fixed fixture produces stable software outputs
- [ ] session split checker rejects overlap
- [ ] normalization uses train split only
- [ ] online-learning-disabled replay leaves weights unchanged
- [ ] bounded online mode records update statistics
- [ ] malformed telemetry fails explicitly instead of silently becoming zero
- [ ] export roundtrip stays compatible with #4 / silicon-bridge contract

## Acceptance criteria

This issue is complete when there is at least one documented supervisor checkpoint with:

1. held-out software replay metrics,
2. deterministic baseline comparison,
3. optional bounded online-adaptation ablation,
4. a defined FPGA parity bundle/contract,
5. no claim of host autonomy without independent safety gating.

## Non-goals

* Live financial trading
* Letting the SNN directly change thermal limits
* Claiming biological nervous-system equivalence
* Cramming a full frontier MoE expert into the Basys 3

## Cross-repo dependencies / follow-ups

* `rmems/Theseus-Quarry` learned-supervisor shadow telemetry
* `rmems/gaming-telemetry` session/replay contract
* `rmems/SynapticDistill.jl` supervisor distillation
* `rmems/silicon-hdl` FPGA parity harness

Suggested labels: `research`, `supervisor`, `neuromorphic`, `fpga`, `online-learning`.

**Comment by rmems** [comment]
Research relationship: #14 now defines the Supervisor v3 feature-engineering ablation. Evaluate identical held-out sessions as RAW (`telemetry → axon-encoder`), KINETIC (`telemetry → kinetic-signals → axon-encoder`), and HYBRID (raw + kinetic features), then report the #6 metrics including false-safe rate, warning lead time, spike rate, latency, and feature overhead. A negative result is acceptable and should be documented.

**Comment by rmems** [comment]
## Current M1 delivery links

- [Spikenaut-SNN #64](https://github.com/rmems/Spikenaut-SNN/pull/64) delivers the completed machine-state anticipation pilot, including held-out continuous temperature/power results against persistence, linear, and memoryless baselines.
- [spikenaut-telemetry-etl #28](https://github.com/rmems/spikenaut-telemetry-etl/pull/28) supplies the audited historical and fresh-preparation input path for that evaluation.

These PRs advance M1 baseline evidence for this issue. They do not close this cross-stage umbrella: policy validation, online adaptation, and FPGA parity remain separate acceptance work.
